src/editor/ui_menus/main_menu_random_map.cc

```diff
@@ -265,72 +306,117 @@ MainMenuNewRandomMap::MainMenuNewRandomMap(EditorInteractive& parent,
 	mountains_box_.add(&mountains_);
 	mountains_box_.set_size(box_width_, mountains_label_.get_h());
 
-	box_.add(&mountains_box_);
-	box_height += margin_ + mountains_box_.get_h();
+	box_.add(&mountains_box_, UI::Box::Resizing::kExpandBoth);
 	box_.add_space(margin_);
-	box_height += margin_;
 
 	// ---------- Island mode ----------
 
 	island_mode_.set_state(true);
-	box_.add(&island_mode_);
-	box_height += margin_ + island_mode_.get_h();
+	box_.add(&island_mode_, UI::Box::Resizing::kExpandBoth);
 	box_.add_space(margin_);
-	box_height += margin_;
 
 	island_mode_.changed.connect([this]() { button_clicked(ButtonId::kIslandMode); });
 
 	// ---------- Random map number edit ----------
 
-	map_number_box_.add(&map_number_label_);
+	map_number_and_id_vbox_1_.add(&map_number_label_, UI::Box::Resizing::kExpandBoth);
 
 	map_number_edit_.changed.connect([this]() { nr_edit_box_changed(); });
 	RNG rng;
 	rng.seed(clock());
 	rng.rand();
 	map_number_ = rng.rand();
 	map_number_edit_.set_text(std::to_string(static_cast<unsigned int>(map_number_)));
-	map_number_box_.add(&map_number_edit_);
-	box_.add(&map_number_box_);
+	map_number_and_id_vbox_2_.add(&map_number_edit_, UI::Box::Resizing::kExpandBoth);
 
-	box_height += margin_ + map_number_box_.get_h();
 	box_.add_space(margin_);
-	box_height += margin_;
 
 	// ---------- Map ID String edit ----------
 
-	map_id_box_.add(&map_id_label_);
+	map_number_and_id_vbox_1_.add(&map_id_label_, UI::Box::Resizing::kExpandBoth);
 
 	map_id_edit_.set_text("abcd-efgh-ijkl-mnop");
 	map_id_edit_.changed.connect([this]() { id_edit_box_changed(); });
-	map_id_box_.add(&map_id_edit_);
-	box_.add(&map_id_box_);
-	box_height += margin_ + map_id_edit_.get_h();
+	map_number_and_id_vbox_2_.add(&map_id_edit_, UI::Box::Resizing::kExpandBoth);
+
+	map_number_and_id_hbox_.add(&map_number_and_id_vbox_1_, UI::Box::Resizing::kExpandBoth);
+	map_number_and_id_hbox_.add(&map_number_and_id_vbox_2_, UI::Box::Resizing::kExpandBoth);
+	box_.add(&map_number_and_id_hbox_, UI::Box::Resizing::kExpandBoth);
 	box_.add_space(margin_);
-	box_height += margin_;
+
+	// ---------- Game-specific stuff ----------
+
+	if (game) {
+		tribe_->add(pgettext("tribe", "Random"), "",
+		            g_image_cache->get("images/ui_fsmenu/random.png"), true,
+		            _("The tribe will be selected at random"));
+
+		bool first_pass = true;
+		for (const Widelands::TribeBasicInfo& tribeinfo : Widelands::get_all_tribeinfos()) {
+			tribe_->add(tribeinfo.descname, tribeinfo.name, g_image_cache->get(tribeinfo.icon), false,
+			            tribeinfo.tooltip);
+
+			// I here assume that all tribes use equivalents of the same starting
```

Comment:
This is fragile. How about assigning the random tribe here already, this way there won't be any trouble.

Or let the random map generation be followed by the normal game setup screen. This way, the player could also set the win conditions and assign the teams freely. I'm pretty sure we'll get a feature request for this anyway after merging this branch.

Reply:
Okay, will do that…